### DISCUSSION TITLE
stream capture issue

### DISCUSSION THREAD

**Discussion by chihualee** [thesofproject/sof#9362] [Q&A]
For stream capture scenario, I am tracing the behavior of Host (sof/src/audio/host-legacy.c and sof/src/drivers/generic/dummy-dma.c).

The data flow should be: Local-buffer => DMA-buffer => buffer-share-linux.

But I found the data flow in Host is:
First copy data from DMA-buffer to buffer-share-linux,
then copy data from Local-buffer to DMA-buffer ?

I think it's not correct. Did I miss something ?

Thanks.

**Comment by dbaluta** [comment]
Are you sure? Can you tell us what board are you using? Can you post some code snippets?

**Reply by chihualee** [reply]
in sof/src/drivers/generic/dummy-dma.c as below,

first invoke **dummy_dma_do_copies**() to **copy data from DMA-buffer to buffer-share-linux**,
then invoke notifier_event(...,NOTIFIER_ID_DMA_COPY, ...) that invoke host_dma_cb() that invoke **host_common_update**() to do 
		source = hd->local_buffer;
		sink = hd->dma_buffer;
		ret = dma_buffer_copy_to(source, sink, hd->process, bytes);
		/* **copy data from Local buffer to DMA-buffer** */

I think the order is incorrect...but it's correct for playback scenario.
		

> static int dummy_dma_copy(struct dma_chan_data *channel, int bytes,
> 			  uint32_t flags)
> {
> 	ssize_t copied;
> 	struct dma_cb_data next = {
> 		.channel = channel,
> 	};
> 	struct dma_chan_pdata *pdata = dma_chan_get_data(channel);
> 
> 	copied = dummy_dma_do_copies(pdata, bytes);
> 	if (copied < 0)
> 		return copied;
> 
> 	next.elem.size = copied;
> 
> 	/* Let the user of the driver know how much we copied */
> 	notifier_event(channel, NOTIFIER_ID_DMA_COPY,
> 		       NOTIFIER_TARGET_CORE_LOCAL, &next, sizeof(next));
> 
> 	return 0;
> }
>

**Comment by lgirdwood** [comment]
@chihualee for capture copying from DSP to host driver, there are two buffers.
1) "Host" buffer in DSP memory. Capture DSP pipeline writes to this and host driver reads from this (host can use DMA or memcpy).
2) ALSA PCM driver buffer in host memory, allocated by driver. Userspace reads from this and driver writes to it (either DMA or memcpy).

The critical thing is to make sure that your R/W pointers are correctly aligned between host driver and DSP FW. i.e. to avoid overrun or underrun. Pls also make sure that the DSP pipeline R/W pointers are updating and buffer 1 is filling, then check same for buffer 2. 

ALSA also needs to some data in the host driver buffer 2 before it will unblock a read() from userspace.

**Reply by chihualee** [reply]
Here, I **ONLY focus the Host component**(src/audio/host-legacy.c and sof/src/drivers/generic/dummy-dma.c) **behavior in DSP side**, regardless of Linux ALSA side.

The DSP pipeline is 
Host <==== DAI

Host have 3 buffers:
1. local buffer: it is between Host and DAI.
2. DMA buffer
3. Host buffer: it is between Host and Linux ALSA.
struct host_data {
...
	struct comp_buffer *dma_buffer;
	struct comp_buffer *local_buffer;
...
	/* local and host DMA buffer info */
	struct hc_buf host;
...
}


In sof/src/drivers/generic/dummy-dma.c
The data transfer order of capture is: **First buffer-2. copy to buffer-3., then buffer-1. copy to buffer-2.**
In playback scenario, it swap the copy-from and copy-to, and it's correct.
But in capture scenario, it's incorrect. Why do buffer-2. copy to buffer-3. even if buffer-1. not copy to buffer-2. yet.

**Reply by lgirdwood** [reply]
Its possible you have a bug in your topology, or in your code if the copy from buffer 1 to 2 is not being performed. You may want to enable logging and look at your logs  to see why copy 1 to 2 is not done.

**Reply by chihualee** [reply]
My SOF version is release v2.8 https://github.com/thesofproject/sof/releases/tag/v2.8.
I don't modify any code in Host component, just print some logs to make sure buffer copy behavior.

As I mentioned before, the SOF native code in Host component **host_copy()**:

**First** invoke **dummy_dma_do_copies()** to copy data from **DMA-buffer to Host-buffer(share with Linux)**,

**Then** then invoke notifier_event(...,NOTIFIER_ID_DMA_COPY, ...) that invoke host_dma_cb() that invoke **host_common_update()** to copy data **from local-buffer to DMA-buffer.**

The copy order seems NOT correct.

Maybe you can **print wPtr rPtr of associated buffers in dummy_dma_do_copies() and host_common_update()**, and you can get what I mean.

**Reply by dbaluta** [reply]
Can you use for reference `main` branch or at least `stable-v10`. Also, what platforms are you using? dummy_dma is not used on Intel platforms at all.

It is used on platforms that can directly share a buffer between ARM core and the DSP. And this is just to emulation Intel DMA behavior.

I wonder why we do not see the issue in our setup. 

The recording should be bad, right?

**Reply by chihualee** [reply]
_Can you use for reference main branch or at least stable-v10._ 
=> https://github.com/thesofproject/sof, 
I use main branch stabel v2.8(https://github.com/thesofproject/sof/releases/tag/v2.8).
There is no stable-v10...

_Also, what platforms are you using? dummy_dma is not used on Intel platforms at all._
=> My platform is running on Xtensa(HiFi).
The pipeline of capture has component "Host" (src/audio/host-legacy.c).
It need to enable DUMMY_DMA manually to build sof/src/drivers/generic/dummy-dma.c, dummy-DMA is used in Host component.

_The recording should be bad, right?_
=> Actually, the recording is correct. 
I think the incorrect copy order ONLY cause the first DMA buffer copy size incorrect.
i.e. if DMA buffer size = 0x780, only the first 0x780 output is incorrect.

**Reply by dbaluta** [reply]
@chihualee  I see your point.  I will need some more time to investigate this. 

If you have any suggestion for patches feel free to send them for review. We are now switching from SOF standalone drivers to Zephyr drivers so we are no longer focusing on sof/src/drivers/generic/dummy-dma.c but we accept patches if needed.

Can you please try this driver from Zephyr:

drivers/dma/dma_nxp_sof_host_dma.c

This should be the replacemnt for dummy-dma.c from SOF.

**Reply by chihualee** [reply]
https://github.com/thesofproject/sof/tree/main/src/drivers

I can't see the file drivers/dma/dma_nxp_sof_host_dma.c you mentioned.

**Reply by chihualee** [reply]
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/dma/dma_nxp_sof_host_dma.c

I got it in Zephyr. But My SOF(2.8) and Zephyr(3.5) version is not too new enough...

Then I try to checkout SOF to the latest version as below.
commit 2e39952d4b7b855b3b85b02c3bf045b15eb467b3
Author: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>
Date:   Mon Aug 19 15:05:39 2024 +0300

Ther I grep dma_nxp_sof_host_dma under SOF folder, and nothing return... I think maybe it's not easy to build code with dma_nxp_sof_host_dma.c....
sof$ grep -r dma_nxp_sof_host_dma .

**Reply by dbaluta** [reply]
Yes, the driver is in zephyr repo. We switched from SOF standalone drivers to Zephyr. 

Is very easy to build latest SOF. 

Just clone the lastest sof version and then run:

$ ./scripts/xtensa-build-zpehyr -u -p imx8m (or whatever platform you use)

**Reply by chihualee** [reply]
@dbaluta 

I tried to fix this bug, however the original design is bundled with capture and playback, it's not easy to fix it robustly.

The attached file is my patch, and I test it work.

Can you review this patch ? (You can disable log in patch manually to prevent log overflow)

[my_patch.patch](https://github.com/user-attachments/files/16757198/my_patch.patch)

**Reply by lgirdwood** [reply]
@chihualee can you send your patch as a pull request for review/merge process. Thanks !
